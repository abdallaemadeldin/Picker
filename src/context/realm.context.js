import React, { createContext, useContext } from "react"
import Realm from "realm";
import { dbOptions, LOG_OBJECT_SCHEME } from "src/db/log.scheme";

export const RealmContext = createContext();

export const RealmProvider = (props) => {

    const insert = values => new Promise((resolve, reject) => {
        Realm.open(dbOptions).then(realm => {
            realm.write(() => {
                realm.create(LOG_OBJECT_SCHEME, values);
                resolve(values);
                console.log("Values added successfully!");
            });
        }).catch(reject)
    })

    const fetch = () => new Promise((resolve, reject) => {
        Realm.open(dbOptions).then(realm => {
            realm.write(() => {
                let all = realm.objects(LOG_OBJECT_SCHEME);
                resolve(all);
            });
        }).catch(reject)
    })

    const contextValue = { insert, fetch }

    return (
        <RealmContext.Provider value={contextValue}>
            {props.children}
        </RealmContext.Provider>
    );
}

export const useRealm = () => useContext(RealmContext);