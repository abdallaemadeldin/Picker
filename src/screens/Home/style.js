import { StyleSheet } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

export const style = () => {
    const { top } = useSafeAreaInsets();
    return StyleSheet.create({
        container: {
            flex: 1,
            backgroundColor: '#ccc',
            justifyContent: 'center',
            alignItems: 'center',
            paddingTop: top + 20
        },
        box: {
            width: '46%',
            margin: '2%',
            height: 130,
            borderRadius: 4,
            backgroundColor: '#fff',
            justifyContent: 'center',
            alignItems: 'center'
        },
        label: {
            fontSize: 24,
            color: '#000',
            fontWeight: 'bold'
        },
        btn: {
            width: '90%',
            marginVertical: 10,
            height: 44,
            backgroundColor: '#fff',
            alignSelf: 'center',
            borderRadius: 4,
            justifyContent: 'center',
            alignItems: 'center',
        },
        btnLabel: {
            fontSize: 16,
            color: '#000',
            fontWeight: 'bold'
        },
        overlay: {
            position: 'absolute',
            top: 0,
            bottom: 0,
            left: 0,
            right: 0,
            backgroundColor: '#0007',
            justifyContent: 'center',
            alignItems: 'center',
        },
        list: {
            width: '100%',
            minHeight: 150,
            maxHeight: '60%',
            backgroundColor: '#fff'
        },
        valuesStyle: {
            fontSize: 20,
            color: '#000',
            fontWeight: 'bold',
            marginHorizontal: 20,
            lineHeight: 24
        }
    });
}