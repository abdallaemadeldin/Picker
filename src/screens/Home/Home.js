import React, { memo } from 'react';
import { View, FlatList, Text, TouchableOpacity, Modal } from 'react-native';
import { useHome } from 'src/hooks';
import { style } from './style';

const Home = () => {
    const { container, box, label, btn, btnLabel, overlay, list, valuesStyle } = style();
    const { values, logs, visible, setVisible, fetchLogs, generate } = useHome();

    const renderItem = ({ item }) => {
        return (
            <View style={box}>
                <Text style={label}>{item}</Text>
            </View>
        );
    }

    const renderFooter = () => {
        return (
            <View style={{ marginTop: 50 }}>
                <TouchableOpacity style={btn} onPress={generate}>
                    <Text style={btnLabel}>Generate</Text>
                </TouchableOpacity>

                <TouchableOpacity style={btn} onPress={fetchLogs}>
                    <Text style={btnLabel}>Show Logs</Text>
                </TouchableOpacity>
            </View>
        );
    }

    return (
        <View style={container}>
            <FlatList
                data={values}
                renderItem={renderItem}
                keyExtractor={(_, i) => i.toString()}
                bounces={false}
                getItemLayout={(_, index) => ({ index, length: 130, offset: 130 * index })}
                numColumns={2}
                centerContent={true}
                ListFooterComponent={renderFooter}
            />

            <Modal visible={visible} animationType="slide" transparent onDismiss={() => setVisible(false)}>
                <View style={{ flex: 1, justifyContent: 'flex-end' }}>
                    <TouchableOpacity style={overlay} onPress={() => setVisible(false)} />
                    <View style={list}>
                        <FlatList
                            data={logs}
                            renderItem={({ item, index }) => {
                                return (
                                    <Text style={valuesStyle} key={index}>{item.values}</Text>
                                );
                            }}
                            bounces={false}
                            keyExtractor={e => e.id}
                        />
                    </View>
                </View>
            </Modal>
        </View>
    );
}

export default memo(Home);