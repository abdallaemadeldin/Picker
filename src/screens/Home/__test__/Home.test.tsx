import React from 'react';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import renderer from 'react-test-renderer';
import { RealmProvider } from 'src/context/realm.context';
import Home from '../Home';

jest.mock("react-redux", () => {
    return {
        useSelector: () => ({
            values: []
        }),
        useDispatch: jest.fn
    }
})

it('renders correctly', () => {
    const component = (
        <SafeAreaProvider initialMetrics={{ insets: { top: 0, left: 0, bottom: 0, right: 0 }, frame: { height: 0, width: 0, x: 0, y: 0 } }}>
            <RealmProvider>
                <Home />
            </RealmProvider>
        </SafeAreaProvider>
    );
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
});