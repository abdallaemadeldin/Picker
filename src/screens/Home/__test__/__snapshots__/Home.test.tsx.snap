// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ccc",
        "flex": 1,
        "justifyContent": "center",
        "paddingTop": 20,
      }
    }
  >
    <RCTScrollView
      ListFooterComponent={[Function]}
      bounces={false}
      centerContent={true}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Object {
                "marginTop": 50,
              }
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "backgroundColor": "#fff",
                  "borderRadius": 4,
                  "height": 44,
                  "justifyContent": "center",
                  "marginVertical": 10,
                  "opacity": 1,
                  "width": "90%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#000",
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                Generate
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "backgroundColor": "#fff",
                  "borderRadius": 4,
                  "height": 44,
                  "justifyContent": "center",
                  "marginVertical": 10,
                  "opacity": 1,
                  "width": "90%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#000",
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                Show Logs
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onDismiss={[Function]}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#0007",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "maxHeight": "60%",
              "minHeight": 150,
              "width": "100%",
            }
          }
        >
          <RCTScrollView
            bounces={false}
          >
            <View />
          </RCTScrollView>
        </View>
      </View>
    </Modal>
  </View>
</RNCSafeAreaProvider>
`;
